### PR TITLE
feat(domains): file domain

### LINKED ISSUE #337: Support for a file domain
Generic file domain for processing a given target could open potential integrations with a variety of tools and reporting.

Notably CVE scanning tools such as Grype & Trivy provide json reporting output formats that Lula could consume and map to applicable controls / other OSCAL artifacts (Such as POAM). In-toto attestations are another example

### Additional Context

Optimally this will be an area that allows for experimentation and testing prior to promoting optimizations of distinct scenarios into new domains.

Scope here should be a single file type (json) - but should support adding additional file type support.

### Issue Deliverables

- [ ] Creation of a `files` domain
- [ ] Ability to specify a file path (local or remote) in a validation
- [ ] Ability to process a `json` file (otherwise error)
- [ ] Ability for each provider to accurately evaluate the data
  - [ ] OPA
  - [ ] Kyverno
- [ ] Docs
- [ ] Tests

### Stretch Goal

- [ ] Add support for yaml files (otherwise create an issue for yaml support)

### PR DESCRIPTION
**Pull request by mildwonkey**
## Description
This PR adds a `files` domain which use's OPA's conftest parser to read in a long list of configuration file types* so that rego validations can be written against the contents. This domain works with OPA and the kyverno provider, though for kyverno (as far as I understand) you need to use the kyverno-json module. 

Domain example:
```yaml
domain:
  type: file
  file-spec:
    filepaths:
    - name: grafana-config
       path: custom.ini
```

The file domain takes an array of files, but the implications are that the output DomainResources map uses the file name as a key, so you need to include the filename in the validation:

```yaml
      package validate
      import rego.v1

      # Default values
      default validate := false
      default msg := "Not evaluated"
      
      validate if {
       check_grafana_config.result
      }
      msg = check_grafana_config.msg
      config := input[grafana-config] # <-- MAP KEY HERE!
      protocol := config.server.protocol

      check_grafana_config = {"result": true, "msg": msg} if {
        protocol == "https"
        msg := "Server protocol is set to https"
      } else = {"result": false, "msg": msg} if {
        protocol == "http"
        msg := "Protocol is http. That's bad, mmkay?"
      }
```

Original PR for reference: https://github.com/defenseunicorns/lula/pull/688

Relates to #337

Note to reviewers: besides all the obvious new files, the major _change_ in here is the introduction of a context.Context that starts from the validate command and flows through to `GetResources()` - the actual place that the file domain needed the information in question. I've added a context key named `LulaValidationWorkDir` (NITPICK FREELY! NAMING THINGS IS HARD!) and I'm using that to store the location that the lula command is run from, so that the file domain can build a relative path (if necessary; it also supports abs paths and network downloads) to location the files to evaluate. I kept it as simple as I could and only added the context where necessary, but I think there's potential to use the command context (and additional context keys) as lula command request contexts throughout the codebase. 

----------------------------
From https://www.conftest.dev/:

```
As of today Conftest supports:

CUE
CycloneDX
Dockerfile
EDN
Environment files (.env)
HCL and HCL2
HOCON
Ignore files (.gitignore, .dockerignore)
INI
JSON
Jsonnet
Property files (.properties)
SPDX
TextProto (Protocol Buffers)
TOML
VCL
XML
YAML
```
--------------------------
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [X] Test, docs, adr added or updated as needed
- [X] [Schema Updates](https://github.com/defenseunicorns/lula/blob/main/docs/community-and-contribution/schema-updates.md) applied
- [X [Contributor Guide Steps](https://github.com/defenseunicorns/lula/blob/main/CONTRIBUTING.md) followed

### CODE REVIEW

File: docs/reference/domains/file-domain.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+      config := input["grafana"]
+      protocol := config.server.protocol
+
+      check_grafana_config = {"result": true, "msg": msg} if {
+        protocol == "https"
+        msg := "Server protocol is set to https"
+      } else = {"result": false, "msg": msg} if {
+        protocol == "http"
+        msg := "Grafana protcol is insecure"
```
Comment (brandtkeller) [inline docs/reference/domains/file-domain.md]:
```suggestion
        msg := "Grafana protocol is insecure"
```

**Review by brandtkeller — CHANGES_REQUESTED** [review]
besides a typo my only request would be an unhappy-path E2E test where the data residing in the associated file does not adhere to the policy. I believe you hinted to this with the https vs http example. I don't believe I saw an unhappy path test.

File: docs/reference/domains/file-domain.md (hunk 1 above)
Comment (mildwonkey) [reply docs/reference/domains/file-domain.md]: Thanks for the test pointer - that also let me realize I left out the kyverno test (the fixture made it in 🙄, not the test)~

**Comment by mildwonkey** [comment]
Added test coverage for kyverno (whoops!), not-satisfied validations, and one for invalid inputs (just an empty file; just proof we don't panic)

**Review by brandtkeller — DISMISSED** [review]
No remaining items worth blocking this from merge. 

@meganwolf0 - would appreciate a second pair of eyes. 

Otherwise two items of opportunity I'll call out:
1. We've been discussing the idea of `E2E` tests becoming more CLI-driven in the near future. I believe some of that structure is still under review but I wouldn't be surprised if we swap these E2E tests in the future.
2. Deciding on the role of entangling templating into our tests in order to reduce the file surface. IE for both pass/fail the same component definitions are used. If we templated the path of the config file - we could probably de-duplicate. Again - not a feature that is ready for use just yet. 

Food for thought, otherwise appreciate this addition!

File: docs/reference/domains/file-domain.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+# File Domain
+The File domain allows for validation of arbitrary file contents. The file domain can evaluate local files and network files. Files are copied to a temporary directory for evaluation and deleted afterwards.
+
+## Specification
+The File domain specification accepts a descriptive name for the file as well as it's path. The names must be unique.
```
Comment (meganwolf0) [inline docs/reference/domains/file-domain.md]:
```suggestion
The File domain specification accepts a descriptive name for the file as well as its path. The names must be unique.
```

File: docs/reference/domains/file-domain.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+domain:
+  type: file
+  file-spec:
+    filepaths:
+    - name: config
+      path: grafana.ini
+```
+
+## Supported File Types
+The file domain use's OPA's [conftest](https://conftest.dev) to parse files into a json-compatible format for validations. ∑Both OPA and kyverno (using [kyverno-json](https://kyverno.github.io/kyverno-json/latest/)) can validate files parsed by the file domain.
```
Comment (meganwolf0) [inline docs/reference/domains/file-domain.md]:
```suggestion
The file domain uses OPA's [conftest](https://conftest.dev) to parse files into a json-compatible format for validations. Both OPA and kyverno (using [kyverno-json](https://kyverno.github.io/kyverno-json/latest/)) can validate files parsed by the file domain.
```

**Review by brandtkeller — DISMISSED** [review]
Deconflicted some recent updates to go mod/sum

**Comment by meganwolf0** [comment]
Not sure if I'm not setting things up right, but still not getting the compose behavior I'd expect with the root directory/modified path in the filepath (attached my test files)
[files_domain.zip](https://github.com/user-attachments/files/17281726/files_domain.zip)

The `comp-def-composed` is the result of `lula t compose -f comp-def.yaml` - I was expecting the filepath.path to be relative to the `comp-def-composed` location. 

Also when I run `validate` I'm still getting the GetResources error: `Error running validation: domain GetResources error: error getting source files: open /Users/meganwolf/Documents/defense-unicorns/lula-compliance-lib/grafana.ini: no such file or directory` so I think this probably needs to be addressed in the underlying composition functionality

**Review by meganwolf0 — CHANGES_REQUESTED** [review]

**Comment by meganwolf0** [comment]
(Ok, I added this comment, then deleted when I thought I tested the wrong branch 😅 but I think I got the right one so still stands)
Here are my test files, happy to hear I'm screwing something up in the set up! 
[files_domain.zip](https://github.com/user-attachments/files/17282215/files_domain.zip)

Basically, when I compose and validate, I'd expect the `filepaths[_].path` to be relative with respect to where the component definition is - since we might have the scenario where we'd have nested component definitions (e.g., `import-component-definitions`) or just remote validations.

**Comment by mildwonkey** [comment]
@meganwolf0  - thanks for the example files. I wasn't accounting for compose at all (specifically the multiple levels of nested file path).  I'll dig into refactoring compose to support the file domain and make sure this is covered in the e2e tests.

**Comment by meganwolf0** [comment]
One more thing, and I think this is really more of a team convo (@brandtkeller) but we've called all the files that contain our `types` in the domains/providers `types.go` (personally I like `spec.go`, feels more descriptive on what's contained therein) but I suppose for consistency sake, should we use `types`? Or should we say going forward we'll use spec and probably create an associated re-naming issue of some of the files in the existing domains/providers to be more descriptive/consistent?

**Comment by mildwonkey** [comment]
That's a good point, I did notice that y'all use naming conventions I'm not used to. I don't object to renaming files (even to "types"), but i do want to say that lula's use of "types" is very unintuitive to me. I found files with actual code implementation in a file just called "types", where I rather assumed I'd find at most an interface definition.

**Comment by mildwonkey** [comment]
> Deconflicted some recent updates to go mod/sum

😓 whoops @brandtkeller thank you but fwiw if you push changes to my PRs I will probably not see the changes before force pushing over them after a rebase

**Comment by mildwonkey** [comment]
I've updated the docs to match our recent conversation: the file domain works with relative file paths but if you're using compose, filepaths have to be relative to the output directory/working directory. (I don't like how I worded it here, and I don't like how I worded it there, so please nitpick!)\

@brandtkeller - @meganwolf0 had a good question in her last comment, about naming the "extra" file in the files domain `types.go` instead of `spec.go`. I'm absolutely willing to rename it if y'all want, though I am personally strongly opposed to the naming convention. thoughts?

**Comment by brandtkeller** [comment]
> @brandtkeller - @meganwolf0 had a good question in her last comment, about naming the "extra" file in the files domain `types.go` instead of `spec.go`. I'm absolutely willing to rename it if y'all want, though I am personally strongly opposed to the naming convention. thoughts?

If you believe this is better convention and current guidance only points at consistency as an argument then I'm onboard with keeping proposed convention as long as we get an issue or some traceable artifact that points to eventually aligning said conventions.

File: docs/reference/domains/file-domain.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+* Property files (.properties)
+* SPDX
+* TextProto (Protocol Buffers)
+* TOML
+* VCL
+* XML
+* YAML
+
+## Validations
+When writing validations against files, the filepath `Nam`e must be included as
```
Comment (brandtkeller) [inline docs/reference/domains/file-domain.md]:
```suggestion
When writing validations against files, the filepath `Name` must be included as
```

**Review by brandtkeller — COMMENTED** [review]
Minor typo otherwise new documentation supports discussion and continuation once defined in #715 and #716

**Review by meganwolf0 — APPROVED** [review]

**Review by brandtkeller — APPROVED** [review]

**Pull request merged**